src/cps/search.cpp

```diff
@@ -71,36 +71,71 @@ namespace cps::search {
         }
 
+        fs::path libdir() {
+            // TODO: libdir needs to be configurable based on the personality,
+            //       and different name schemes.
+            //       This is complicated by the fact that different distros have
+            //       different schemes.
+            return "lib";
+        }
+
+        fs::path datadir() {
+            // TODO: needs to be configurable. see libdir above.
+            return "share";
+        }
+
         const std::vector<fs::path> nix{"/usr", "/usr/local"};
         // TODO: const std::vector<std::string> mac{""};
         // TODO: const std::vector<std::string> win{""};
 
         std::vector<fs::path> cached_paths{};
 
-        const std::vector<fs::path> search_paths(Env env) {
+        /// @brief expands a single search prefix into a set of full paths
+        /// @param env stored environment variables
+        /// @return A vector of paths to search, in order
+        std::vector<fs::path> expand_prefix(const fs::path & path) {
+            std::vector<fs::path> paths{};
+
+            // TODO: macOS specific paths
+            // TODO: Windows specific paths
+
+            // TODO: handle name-like search paths
+            paths.emplace_back(path / libdir() / "cps");
+            paths.emplace_back(path / datadir() / "cps");
+
+            return paths;
+        };
+
+        /// @brief Expands CPS search prefixes into concrete paths
+        /// @param env stored environment variables
+        /// @return A vector of paths to search, in order
+        const std::vector<fs::path> search_paths(const Env & env) {
             if (!cached_paths.empty()) {
                 return cached_paths;
             }
 
             if (env.cps_path) {
-                cached_paths.reserve(nix.size());
-                cached_paths.insert(cached_paths.end(), nix.begin(), nix.end());
                 auto && paths = utils::split(env.cps_path.value());
                 cached_paths.reserve(paths.size());
                 cached_paths.insert(cached_paths.end(), paths.begin(), paths.end());
-            } else {
-                cached_paths = nix;
             }
 
-            return cached_paths;
-        }
+            if (env.cps_prefix_path) {
+                auto && prefixes = utils::split(env.cps_prefix_path.value());
+                for (auto && p : prefixes) {
+                    auto && paths = expand_prefix(p);
+                    cached_paths.reserve(cached_paths.size() + paths.size());
+                    cached_paths.insert(cached_paths.end(), paths.begin(), paths.end());
+                }
+            }
 
-        const fs::path libdir() {
-            // TODO: libdir needs to be configurable based on the personality,
-            //       and different name schemes.
-            //       This is complicated by the fact that different distros have
-            //       different schemes.
-            return "lib";
+            for (auto && p : nix) {
```

Comment:
I was confused by the name `nix`, for what it's worth. After poking around, it looks like that's intended to represent the installation prefix for CPS files on Linux and Unix. Maybe call it "nix_prefix" or "system_prefix" or something?

Reply:
I don't forsee the `nix` variable sticking around. Just looking at the specification for Windows and macOS, I don't think it's really a viable solution, since they don't have same "constant prefix, mutable suffix" scheme that *nix does.